vnext/JSI/Shared/ChakraCoreRuntime.h

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#ifdef CHAKRACORE
+
+#include "ChakraCoreRuntimeArgs.h"
+#include "ChakraRuntime.h"
+
+#include "ChakraCoreDebugger.h"
+
+#include <memory>
+#include <string>
+
+namespace Microsoft::JSI {
+
+class ChakraCoreRuntime : public ChakraRuntime {
+ public:
+  ChakraCoreRuntime(const std::shared_ptr<ChakraCoreRuntimeArgs> &args) noexcept;
+  ~ChakraCoreRuntime() noexcept;
+
+#pragma region Functions_inherited_from_Runtime
+
+  facebook::jsi::Value evaluatePreparedJavaScript(
+      const std::shared_ptr<const facebook::jsi::PreparedJavaScript> &preparedJs) override;
+
+  std::string description() override;
+
+ private:
+  // We override the following because ChakraCore (unlike Chakra) offers
+  // weak reference semantics.
+  facebook::jsi::WeakObject createWeakObject(const facebook::jsi::Object &obj) override;
+  facebook::jsi::Value lockWeakObject(const facebook::jsi::WeakObject &weakObj) override;
+
+#pragma endregion Functions_inherited_from_Runtime
+
+#pragma region Functions_inherited_from_ChakraRuntime
+
+  facebook::jsi::Value EvaluateJavaScriptSimple(
+      const std::shared_ptr<const facebook::jsi::Buffer> &buffer,
+      const std::string &sourceUrl) override;
+
+  std::shared_ptr<const facebook::jsi::Buffer> SerializeScript(
+      const std::shared_ptr<const facebook::jsi::Buffer> &buffer) override;
+
+  void StartDebugging() override;
+
+  void InitializeAdditionalFeatures() override;
+
+#pragma endregion Functions_inherited_from_ChakraRuntime
+
+  using WeakChakraPointerValue = ChakraPointerValueTemplate<JsWeakRef>;
+
+  // source must be a JS ArrayBuffer or JS ExternalArrayBuffer. sourceUrl must
+  // be a JS string.
+  facebook::jsi::Value EvaluateJavaScriptImpl(const ChakraObjectRef &source, const ChakraObjectRef &sourceUrl);
+
+  // In-proc debugging helpers
+  void ProcessDebuggerCommandQueue();
+  static void CALLBACK ProcessDebuggerCommandQueueCallback(void *callbackState);
```

Comment:
>CALLBACK [](start = 14, length = 8)

OOC why is this needed?

Reply:
We need to pass a callback of this signature to ChakraCoreDebugger API when setting up direct debugging.

---
In reply to: [345520533](https://github.com/microsoft/react-native-windows/pull/3650#discussion_r345520533) [](ancestors = 345520533)